components/resizable/nz-resizable.directive.ts

```diff
@@ -0,0 +1,325 @@
+    '(mouseenter)': 'onMouseenter()',
+    '(mouseleave)': 'onMouseleave()'
+  }
+})
+export class NzResizableDirective implements AfterViewInit, OnDestroy {
+  @Input() nzBounds: 'window' | 'parent' | ElementRef<HTMLElement> = 'parent';
+  @Input() nzMaxHeight: number;
+  @Input() nzMaxWidth: number;
+  @Input() nzMinHeight: number = 40;
+  @Input() nzMinWidth: number = 40;
+  @Input() nzGridColumnCount: number = -1;
+  @Input() nzMaxColumn: number = -1;
+  @Input() nzMinColumn: number = -1;
+  @Input() @InputBoolean() nzLockAspectRatio: boolean = false;
+  @Input() @InputBoolean() nzPreview: boolean = false;
+  @Output() readonly nzResize = new EventEmitter<NzResizeEvent>();
+  @Output() readonly nzResizeEnd = new EventEmitter<NzResizeEvent>();
+  @Output() readonly nzResizeStart = new EventEmitter<NzResizeEvent>();
+
+  resizing = false;
+  elRect: ClientRect | DOMRect;
+  currentHandleEvent: NzResizeHandleMouseDownEvent;
+  ghostElement: HTMLDivElement | null;
+  el: HTMLElement;
+  sizeCache: NzResizeEvent | null;
+  private destroy$ = new Subject<void>();
+
+  constructor(
+    private elementRef: ElementRef<HTMLElement>,
+    private renderer: Renderer2,
+    private ngZone: NgZone,
+    private nzResizableService: NzResizableService,
+    private platform: Platform
+  ) {
+    this.nzResizableService.handleMouseDown$.pipe(takeUntil(this.destroy$)).subscribe(event => {
+      this.resizing = true;
+      this.currentHandleEvent = event;
+      this.setCursor();
+      this.nzResizeStart.emit({
+        mouseEvent: event.mouseEvent
+      });
+      this.elRect = this.el.getBoundingClientRect();
+    });
+  }
+
+  @HostListener('document:mouseup', ['$event'])
+  onMouseup($event: MouseEvent): void {
+    this.ngZone.runOutsideAngular(() => {
+      if (this.resizing) {
+        this.resizing = false;
+        this.nzResizableService.documentMouseUp$.next();
+        this.endResize($event);
+      }
+    });
+  }
+
+  @HostListener('document:mousemove', ['$event'])
+  onMousemove($event: MouseEvent): void {
+    this.ngZone.runOutsideAngular(() => {
+      if (this.resizing) {
+        this.resize($event);
+      }
+    });
+  }
+
+  onMouseenter(): void {
+    this.nzResizableService.mouseEntered$.next(true);
+  }
+
+  onMouseleave(): void {
+    this.nzResizableService.mouseEntered$.next(false);
+  }
+
+  setPosition(): void {
+    const position = getComputedStyle(this.el).position;
+    if (position === 'static' || !position) {
+      this.renderer.setStyle(this.el, 'position', 'relative');
+    }
+  }
+
+  calcSize(width: number, height: number, ratio: number): NzResizeEvent {
+    let newWidth: number;
+    let newHeight: number;
+    let col = 0;
+    let spanWidth = 0;
+    let maxWidth = Infinity;
+    let minWidth = this.nzMinWidth;
+    let maxHeight = Infinity;
+    if (this.nzBounds === 'parent') {
+      const parent = this.renderer.parentNode(this.el);
+      if (parent instanceof HTMLElement) {
```

Comment:
I guess these three algorithms are the same? There should be a function like `ensureInBoundary` for simplicity.